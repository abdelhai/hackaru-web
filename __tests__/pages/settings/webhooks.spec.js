import { Store } from 'vuex-mock-store';
import Factory from '@/__tests__/__setups__/factory';
import Webhooks from '@/pages/settings/webhooks';

describe('Webhooks', () => {
  let factory;
  let wrapper;

  const $store = new Store({
    getters: {
      'webhooks/getWebhooks': [
        {
          id: 1,
          userId: 1,
          targetUrl: 'http://www.example.com/1',
          event: 'activity:updated',
          createdAt: '2019-02-06T10:23:24.199Z',
          updatedAt: '2019-02-06T10:23:24.199Z'
        },
        {
          id: 2,
          userId: 1,
          targetUrl: 'http://www.example.com/2',
          event: 'activity:created',
          createdAt: '2019-02-06T10:23:24.199Z',
          updatedAt: '2019-02-06T10:23:24.199Z'
        }
      ]
    }
  });

  beforeEach(() => {
    $store.reset();
    factory = new Factory(Webhooks, {
      mocks: { $store }
    });
  });

  it('render correctly', () => {
    expect(factory.shallow().element).toMatchSnapshot();
  });

  describe('when click delete button and confirm is true', () => {
    beforeEach(() => {
      global.confirm = () => true;
      wrapper = factory.shallow();
      wrapper
        .findAll('.delete-button')
        .at(1)
        .vm.$emit('click');
    });

    it('dispatch webhooks/deleteWebhook', () => {
      expect($store.dispatch).toHaveBeenCalledWith('webhooks/deleteWebhook', 2);
    });

    it('send ga event', () => {
      expect(factory.options.mocks.$ga.event).toHaveBeenCalledWith(
        'webhook',
        'deleteWebhook'
      );
    });
  });

  describe('when click delete button and confirm is false', () => {
    beforeEach(() => {
      global.confirm = () => false;
      wrapper = factory.shallow();
      wrapper
        .findAll('.delete-button')
        .at(1)
        .vm.$emit('click');
    });

    it('does not dispatch webhooks/deleteWebhook', () => {
      expect($store.dispatch).not.toHaveBeenCalledWith(
        'webhooks/deleteWebhook',
        2
      );
    });
  });

  describe('when click add button', () => {
    beforeEach(() => {
      wrapper = factory.shallow();
      wrapper.find('.add-button').vm.$emit('click');
    });

    it('render modal correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('show modal', () => {
      expect(factory.options.mocks.$modal.show).toHaveBeenCalledWith('webhook');
    });
  });

  describe('when submit create webhook form', () => {
    beforeEach(() => {
      $store.dispatch.mockReturnValue(true);
      wrapper = factory.shallow();
      wrapper.setData({
        event: 'activity:created',
        targetUrl: 'http://example.com'
      });
      wrapper.find('form').trigger('submit.prevent');
    });

    it('dispatch webhooks/addWebhook', () => {
      expect($store.dispatch).toHaveBeenCalledWith('webhooks/addWebhook', {
        event: 'activity:created',
        targetUrl: 'http://example.com'
      });
    });

    it('send ga event', () => {
      expect(factory.options.mocks.$ga.event).toHaveBeenCalledWith(
        'webhook',
        'addWebhook'
      );
    });

    it('hide modal', () => {
      expect(factory.options.mocks.$modal.hide).toHaveBeenCalledWith('webhook');
    });
  });

  describe('when submit create webhook form but failed', () => {
    beforeEach(() => {
      $store.dispatch.mockReturnValue(false);
      wrapper = factory.shallow();
      wrapper.find('form').trigger('submit.prevent');
    });

    it('does not hide modal', () => {
      expect(factory.options.mocks.$modal.hide).not.toHaveBeenCalled();
    });
  });
});
