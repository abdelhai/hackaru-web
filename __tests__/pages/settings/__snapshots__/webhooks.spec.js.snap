// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhooks render correctly 1`] = `
<div
  class="settings"
>
  <modal-wrapper-stub
    height="auto"
    name="webhook"
  >
    <form>
      <modal-header-stub>
        <h1>
          modalTitle
        </h1>
      </modal-header-stub>
       
      <modal-item-stub>
        <modal-label-stub>
          
          event
        
        </modal-label-stub>
         
        <select>
          <option
            value="activity:created"
          >
            
            events.activity:created
          
          </option>
          <option
            value="activity:stopped"
          >
            
            events.activity:stopped
          
          </option>
          <option
            value="activity:updated"
          >
            
            events.activity:updated
          
          </option>
          <option
            value="activity:deleted"
          >
            
            events.activity:deleted
          
          </option>
          <option
            value="project:created"
          >
            
            events.project:created
          
          </option>
          <option
            value="project:updated"
          >
            
            events.project:updated
          
          </option>
          <option
            value="project:deleted"
          >
            
            events.project:deleted
          
          </option>
        </select>
      </modal-item-stub>
       
      <modal-item-stub>
        <modal-label-stub>
          
          targetUrl
        
        </modal-label-stub>
         
        <input
          class="input inline"
          placeholder="https://"
          type="url"
        />
      </modal-item-stub>
       
      <modal-footer-stub>
        <btn-stub
          class="is-rounded is-primary"
          type="submit"
        >
          
          add
        
        </btn-stub>
      </modal-footer-stub>
    </form>
  </modal-wrapper-stub>
   
  <section
    class="content"
  >
    <header
      class="header"
    >
      <heading-stub
        class="is-small"
      >
        <icon-stub
          name="anchor-icon"
        />
        
        title
      
      </heading-stub>
       
      <btn-stub
        aria-label="ariaLabels.add"
        class="is-primary is-circle has-dropshadow add-button"
        type="button"
      >
        <icon-stub
          name="plus-icon"
        />
      </btn-stub>
    </header>
     
    <div
      class="list-item"
    >
      <p>
        events.activity:updated
      </p>
       
      <h1>
        http://www.example.com/1
      </h1>
       
      <btn-stub
        aria-label="ariaLabels.delete"
        class="delete-button has-icon"
        type="button"
      >
        <icon-stub
          class="is-danger"
          name="x-icon"
        />
      </btn-stub>
    </div>
    <div
      class="list-item"
    >
      <p>
        events.activity:created
      </p>
       
      <h1>
        http://www.example.com/2
      </h1>
       
      <btn-stub
        aria-label="ariaLabels.delete"
        class="delete-button has-icon"
        type="button"
      >
        <icon-stub
          class="is-danger"
          name="x-icon"
        />
      </btn-stub>
    </div>
     
    <!---->
  </section>
</div>
`;

exports[`Webhooks when click add button render modal correctly 1`] = `
<div
  class="settings"
>
  <modal-wrapper-stub
    height="auto"
    name="webhook"
  >
    <form>
      <modal-header-stub>
        <h1>
          modalTitle
        </h1>
      </modal-header-stub>
       
      <modal-item-stub>
        <modal-label-stub>
          
          event
        
        </modal-label-stub>
         
        <select>
          <option
            value="activity:created"
          >
            
            events.activity:created
          
          </option>
          <option
            value="activity:stopped"
          >
            
            events.activity:stopped
          
          </option>
          <option
            value="activity:updated"
          >
            
            events.activity:updated
          
          </option>
          <option
            value="activity:deleted"
          >
            
            events.activity:deleted
          
          </option>
          <option
            value="project:created"
          >
            
            events.project:created
          
          </option>
          <option
            value="project:updated"
          >
            
            events.project:updated
          
          </option>
          <option
            value="project:deleted"
          >
            
            events.project:deleted
          
          </option>
        </select>
      </modal-item-stub>
       
      <modal-item-stub>
        <modal-label-stub>
          
          targetUrl
        
        </modal-label-stub>
         
        <input
          class="input inline"
          placeholder="https://"
          type="url"
        />
      </modal-item-stub>
       
      <modal-footer-stub>
        <btn-stub
          class="is-rounded is-primary"
          type="submit"
        >
          
          add
        
        </btn-stub>
      </modal-footer-stub>
    </form>
  </modal-wrapper-stub>
   
  <section
    class="content"
  >
    <header
      class="header"
    >
      <heading-stub
        class="is-small"
      >
        <icon-stub
          name="anchor-icon"
        />
        
        title
      
      </heading-stub>
       
      <btn-stub
        aria-label="ariaLabels.add"
        class="is-primary is-circle has-dropshadow add-button"
        type="button"
      >
        <icon-stub
          name="plus-icon"
        />
      </btn-stub>
    </header>
     
    <div
      class="list-item"
    >
      <p>
        events.activity:updated
      </p>
       
      <h1>
        http://www.example.com/1
      </h1>
       
      <btn-stub
        aria-label="ariaLabels.delete"
        class="delete-button has-icon"
        type="button"
      >
        <icon-stub
          class="is-danger"
          name="x-icon"
        />
      </btn-stub>
    </div>
    <div
      class="list-item"
    >
      <p>
        events.activity:created
      </p>
       
      <h1>
        http://www.example.com/2
      </h1>
       
      <btn-stub
        aria-label="ariaLabels.delete"
        class="delete-button has-icon"
        type="button"
      >
        <icon-stub
          class="is-danger"
          name="x-icon"
        />
      </btn-stub>
    </div>
     
    <!---->
  </section>
</div>
`;
